cdm/misc/src/main/java/ucar/nc2/geotiff/GeotiffWriter.java

```diff
@@ -144,20 +210,29 @@ void writeGrid(GridDatatype grid, Array data, boolean greyScale, double xStart,
       ArrayByte result = replaceMissingValuesAndScale(grid, data, dataMinMax);
       nextStart = geotiff.writeData((byte[]) result.getStorage(), imageNumber);
     } else {
-      ArrayFloat result = replaceMissingValues(grid, data, dataMinMax);
-      nextStart = geotiff.writeData((float[]) result.getStorage(), imageNumber);
+      if (dtype == DataType.BYTE || dtype == DataType.UBYTE) {
+        ArrayByte result = coerceByte(data);
+        nextStart = geotiff.writeData((byte[]) result.getStorage(), imageNumber);
+      } else if (dtype.isIntegral()) {
+        ArrayInt result = coerceInt(data);
+        nextStart = geotiff.writeData((int[]) result.getStorage(), imageNumber);
+      } else {
+        ArrayFloat result = replaceMissingValues(grid, data, dataMinMax);
+        nextStart = geotiff.writeData((float[]) result.getStorage(), imageNumber);
+      }
```

Comment:
I should see what I can do about code deduplication, as this is nearly identical to code in another `writeGrid()` method.

Reply:
Should also see about adding some unit tests to make sure that the data is getting saved correctly in different formats and coerced correctly.